File: src/Exb.Mobile.Shared/Services/ApiClient.cs
using System.Net;
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Serialization;
using Exb.Mobile.Shared.Models;

namespace Exb.Mobile.Shared.Services;

public sealed class ApiException(string message, int? statusCode = null, bool isNetwork = false) : Exception(message)
{
    public int? StatusCode { get; } = statusCode;
    public bool IsNetwork { get; } = isNetwork;
    public bool IsUnauthorised => StatusCode == 401;
}

/// <summary>
/// A thin, hand-written REST client for /api/v1 — mirrors the original Flutter
/// app's ApiClient deliberately (see the porting notes): no heavy HTTP
/// framework, one JSON shape per endpoint, and error messages meant to be
/// shown to the visitor directly rather than a generic "something went wrong."
/// </summary>
public sealed class ApiClient
{
    private static readonly JsonSerializerOptions Json = new(JsonSerializerDefaults.Web);

    private readonly HttpClient _http;
    private string? _token;

    public event Action? Unauthorised;

    public ApiClient(HttpClient http)
    {
        _http = http;
        _http.Timeout = TimeSpan.FromSeconds(20);
    }

    public string? Token
    {
        get => _token;
        set => _token = value;
    }

    public bool HasToken => !string.IsNullOrEmpty(_token);

    private HttpRequestMessage Build(HttpMethod method, string path, object? body = null, Dictionary<string, string?>? query = null)
    {
        string url = "api/v1" + path;
        if (query is { Count: > 0 })
        {
            var pairs = query.Where(kv => !string.IsNullOrEmpty(kv.Value)).Select(kv => $"{Uri.EscapeDataString(kv.Key)}={Uri.EscapeDataString(kv.Value!)}");
            string qs = string.Join("&", pairs);
            if (qs.Length > 0) url += "?" + qs;
        }

        var request = new HttpRequestMessage(method, url);
        request.Headers.Accept.ParseAdd("application/json");
        if (!string.IsNullOrEmpty(_token))
            request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", _token);
        if (body is not null)
            request.Content = JsonContent.Create(body, options: Json);
        return request;
    }

    private async Task<T> Send<T>(HttpRequestMessage request, CancellationToken ct)
    {
        HttpResponseMessage response;
        try
        {
            response = await _http.SendAsync(request, ct);
        }
        catch (TaskCanceledException) when (!ct.IsCancellationRequested)
        {
            throw new ApiException("The exhibition system is not responding. The network in the hall may be busy — try again in a moment.", isNetwork: true);
        }
        catch (HttpRequestException ex)
        {
            throw new ApiException($"Cannot reach the exhibition system. Check that you are on the venue wifi and try again. ({ex.Message})", isNetwork: true);
        }

        string raw = await response.Content.ReadAsStringAsync(ct);

        if (!response.IsSuccessStatusCode)
        {
            if (response.StatusCode == HttpStatusCode.Unauthorized) Unauthorised?.Invoke();

            string? serverMessage = null;
            try
            {
                using var doc = JsonDocument.Parse(string.IsNullOrWhiteSpace(raw) ? "{}" : raw);
                if (doc.RootElement.TryGetProperty("error", out var err) && err.ValueKind == JsonValueKind.String)
                    serverMessage = err.GetString();
            }
            catch (JsonException) { /* malformed body — fall through to the status-keyed message */ }

            throw new ApiException(serverMessage ?? FallbackMessage((int)response.StatusCode), (int)response.StatusCode);
        }

        if (string.IsNullOrWhiteSpace(raw)) return default!;
        try
        {
            return JsonSerializer.Deserialize<T>(raw, Json)!;
        }
        catch (JsonException)
        {
            return default!;
        }
    }

    private static string FallbackMessage(int status) => status switch
    {
        400 => "That request was not understood by the exhibition system.",
        401 => "Please sign in again.",
        404 => "That is no longer available.",
        429 => "Too many attempts. Wait a moment and try again.",
        >= 500 => "The exhibition system had a problem. Try again shortly.",
        _ => $"Something went wrong (error {status}).",
    };

    // --- auth -----------------------------------------------------------

    public Task<LoginCodeRequest> RequestLoginCode(string email, CancellationToken ct = default)
        => Send<LoginCodeRequest>(Build(HttpMethod.Post, "/auth/request-code", new { email }), ct);

    public async Task<(string Token, Visitor Visitor)> VerifyLoginCode(
        string email, string code, string? platform, string? deviceName, string? appVersion, CancellationToken ct = default)
    {
        var result = await Send<VerifyResponse>(Build(HttpMethod.Post, "/auth/verify",
            new { email, code, platform, deviceName, appVersion }), ct);

        if (string.IsNullOrEmpty(result.Token))
            throw new ApiException("Sign-in did not return a valid session. Try again.");

        _token = result.Token;
        return (result.Token, result.Visitor);
    }

    public async Task Logout(CancellationToken ct = default)
    {
        try { await Send<object>(Build(HttpMethod.Post, "/auth/logout"), ct); }
        catch (ApiException) { /* best-effort — the token is cleared locally regardless */ }
        finally { _token = null; }
    }

    // --- visitor's own profile -------------------------------------------

    public Task<Visitor> Me(CancellationToken ct = default)
        => Send<Visitor>(Build(HttpMethod.Get, "/me"), ct);

    public Task<Visitor> UpdateConsent(bool? email = null, bool? tracking = null, CancellationToken ct = default)
        => Send<Visitor>(Build(HttpMethod.Patch, "/me/consent", new { consentEmail = email, consentTracking = tracking }), ct);

    // --- exhibition bootstrap / directory ---------------------------------

    public Task<Exhibition> GetExhibition(CancellationToken ct = default)
        => Send<Exhibition>(Build(HttpMethod.Get, "/exhibition"), ct);

    public Task<List<Hall>> Halls(CancellationToken ct = default)
        => Send<List<Hall>>(Build(HttpMethod.Get, "/halls"), ct);

    public Task<HallDetail> Hall(int id, int page = 1, int pageSize = 50, CancellationToken ct = default)
        => Send<HallDetail>(Build(HttpMethod.Get, $"/halls/{id}", query: new()
        {
            ["page"] = page.ToString(),
            ["pageSize"] = pageSize.ToString(),
        }), ct);

    public Task<List<Category>> Categories(CancellationToken ct = default)
        => Send<List<Category>>(Build(HttpMethod.Get, "/categories"), ct);

    public Task<Paged<Exhibitor>> Exhibitors(
        string? query = null, int? categoryId = null, int? subCategoryId = null, int? hallId = null, string? country = null,
        int page = 1, int pageSize = 25, CancellationToken ct = default)
        => Send<Paged<Exhibitor>>(Build(HttpMethod.Get, "/exhibitors", query: new()
        {
            ["q"] = query,
            ["categoryId"] = categoryId?.ToString(),
            ["subCategoryId"] = subCategoryId?.ToString(),
            ["hallId"] = hallId?.ToString(),
            ["country"] = country,
            ["page"] = page.ToString(),
            ["pageSize"] = pageSize.ToString(),
        }), ct);

    public Task<ExhibitorDetail> Exhibitor(int id, CancellationToken ct = default)
        => Send<ExhibitorDetail>(Build(HttpMethod.Get, $"/exhibitors/{id}"), ct);

    public Task<SearchResults> SearchEverything(string query, CancellationToken ct = default)
        => Send<SearchResults>(Build(HttpMethod.Get, "/search", query: new() { ["q"] = query }), ct);

    // --- programme ----------------------------------------------------------

    public Task<Paged<Session>> Sessions(
        string? query = null, DateOnly? date = null, string? kind = null, int? hallId = null,
        int? categoryId = null, int? subCategoryId = null, bool bookmarkedOnly = false,
        int page = 1, int pageSize = 50, CancellationToken ct = default)
        => Send<Paged<Session>>(Build(HttpMethod.Get, "/sessions", query: new()
        {
            ["q"] = query,
            ["date"] = date?.ToString("yyyy-MM-dd"),
            ["kind"] = kind,
            ["hallId"] = hallId?.ToString(),
            ["categoryId"] = categoryId?.ToString(),
            ["subCategoryId"] = subCategoryId?.ToString(),
            ["bookmarked"] = bookmarkedOnly ? "true" : null,
            ["page"] = page.ToString(),
            ["pageSize"] = pageSize.ToString(),
        }), ct);

    public Task<SessionDetail> Session(int id, CancellationToken ct = default)
        => Send<SessionDetail>(Build(HttpMethod.Get, $"/sessions/{id}"), ct);

    public Task SetBookmarked(int sessionId, bool bookmarked, CancellationToken ct = default)
        => Send<object>(Build(bookmarked ? HttpMethod.Post : HttpMethod.Delete, $"/sessions/{sessionId}/bookmark"), ct);

    // --- scanning and the e-catalogue pack -----------------------------------

    public Task<ScanResult> Scan(string scanned, CancellationToken ct = default)
        => Send<ScanResult>(Build(HttpMethod.Post, "/me/scan", new { token = scanned }), ct);

    public async Task<List<ScannedStand>> MyCatalogues(CancellationToken ct = default)
    {
        var wire = await Send<CataloguesResponse>(Build(HttpMethod.Get, "/me/catalogues"), ct);
        return wire.Items;
    }

    public async Task<int> RequestCatalogue(int kioskId, CancellationToken ct = default)
    {
        var wire = await Send<ScanResult>(Build(HttpMethod.Post, "/me/catalogues", new { kioskId }), ct);
        return wire.TodayCount;
    }

    public Task SetCatalogueIncluded(int kioskId, bool included, CancellationToken ct = default)
        => Send<object>(Build(HttpMethod.Patch, $"/me/catalogues/{kioskId}", new { included }), ct);

    public async Task<VisitorDay> MyDay(CancellationToken ct = default)
    {
        var wire = await Send<VisitorDayWire>(Build(HttpMethod.Get, "/me/day"), ct);
        if (!wire.TrackingConsent)
            return new VisitorDay { TrackingConsent = false, Message = wire.Message };

        var day = wire.Day ?? new VisitorDayPayload();
        return new VisitorDay
        {
            TrackingConsent = true,
            TotalDwellText = day.TotalDwellText,
            StandsWithInterest = day.StandsWithInterest,
            Visited = day.Visited,
            Categories = day.Categories,
            Missed = day.Missed,
        };
    }

    // --- private wire-only shapes (unwrapped into public models above) ------

    private sealed class VerifyResponse
    {
        public string Token { get; set; } = "";
        public Visitor Visitor { get; set; } = new();
    }

    private sealed class CataloguesResponse
    {
        public List<ScannedStand> Items { get; set; } = [];
    }

    private sealed class VisitorDayWire
    {
        public bool TrackingConsent { get; set; }
        public string? Message { get; set; }
        public VisitorDayPayload? Day { get; set; }
    }

    private sealed class VisitorDayPayload
    {
        public string? TotalDwellText { get; set; }
        public int? StandsWithInterest { get; set; }
        public List<VisitedStand> Visited { get; set; } = [];
        public List<CategoryInterest> Categories { get; set; } = [];
        public List<MissedStand> Missed { get; set; } = [];
    }
}

File: MOBILE APP/exhibition_companion/lib/api/api_client.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;

import 'models.dart';

/// Anything the app should show the visitor rather than crash on.
///
/// The message is always something a person standing in a hall can act on —
/// "the code has expired, ask for a new one" — never a status code. The
/// server's own wording is preferred when it sends one, because it knows more
/// about why than we do.
class ApiException implements Exception {
  ApiException(this.message, {this.statusCode, this.isNetwork = false});

  final String message;
  final int? statusCode;
  final bool isNetwork;

  /// The token is gone or revoked: the app must return to the sign-in screen.
  bool get isUnauthorised => statusCode == 401;

  @override
  String toString() => message;
}

/// The typed wrapper around the exhibition's mobile API.
///
/// One instance is shared for the life of the app. It holds the bearer token
/// and nothing else — no caching, because a visitor standing at a stand needs
/// the truth about that stand, not a copy from twenty minutes ago, and the
/// payloads are small enough that refetching costs less than getting it wrong.
class ApiClient {
  ApiClient({required String baseUrl, http.Client? httpClient})
      : _baseUrl = _normalise(baseUrl),
        _http = httpClient ?? http.Client();

  final String _baseUrl;
  final http.Client _http;

  String? _token;

  /// Fired when the server rejects our token, so the app can sign out cleanly
  /// from wherever it happens to be rather than each screen handling it.
  void Function()? onUnauthorised;

  String get baseUrl => _baseUrl;
  bool get hasToken => _token != null;

  set token(String? value) => _token = value;

  static String _normalise(String url) {
    var trimmed = url.trim();
    while (trimmed.endsWith('/')) {
      trimmed = trimmed.substring(0, trimmed.length - 1);
    }
    return trimmed;
  }

  /// Conference wifi is the worst network most apps ever meet, so the timeout
  /// is generous enough to survive a slow hall and short enough that a dead
  /// connection is reported rather than spun on.
  static const Duration _timeout = Duration(seconds: 20);

  Map<String, String> get _headers => {
        'Accept': 'application/json',
        'Content-Type': 'application/json',
        if (_token != null) 'Authorization': 'Bearer $_token',
      };

  Uri _uri(String path, [Map<String, dynamic>? query]) {
    final cleaned = <String, String>{};
    query?.forEach((key, value) {
      if (value == null) return;
      final text = value.toString();
      if (text.isEmpty) return;
      cleaned[key] = text;
    });

    return Uri.parse('$_baseUrl/api/v1$path')
        .replace(queryParameters: cleaned.isEmpty ? null : cleaned);
  }

  Future<dynamic> _send(
    String method,
    String path, {
    Map<String, dynamic>? query,
    Object? body,
  }) async {
    final uri = _uri(path, query);

    try {
      final request = http.Request(method, uri)..headers.addAll(_headers);
      if (body != null) request.body = jsonEncode(body);

      final streamed = await _http.send(request).timeout(_timeout);
      final response = await http.Response.fromStream(streamed);

      return _decode(response);
    } on ApiException {
      rethrow;
    } on SocketException {
      throw ApiException(
        'Cannot reach the exhibition system. Check that you are on the venue '
        'wifi and try again.',
        isNetwork: true,
      );
    } on TimeoutException {
      throw ApiException(
        'The exhibition system is not responding. The network in the hall may '
        'be busy — try again in a moment.',
        isNetwork: true,
      );
    } on http.ClientException catch (e) {
      throw ApiException('Connection problem: ${e.message}', isNetwork: true);
    }
  }

  dynamic _decode(http.Response response) {
    final status = response.statusCode;

    if (status == 401) {
      // Let the app tear the session down once, centrally, rather than in every
      // screen that happens to make the unlucky request.
      onUnauthorised?.call();
    }

    dynamic parsed;
    if (response.body.isNotEmpty) {
      try {
        parsed = jsonDecode(utf8.decode(response.bodyBytes));
      } on FormatException {
        parsed = null;
      }
    }

    if (status >= 200 && status < 300) return parsed;

    final serverSaid =
        parsed is Map<String, dynamic> ? parsed['error']?.toString() : null;

    throw ApiException(
      serverSaid ?? _fallbackMessage(status),
      statusCode: status,
    );
  }

  String _fallbackMessage(int status) => switch (status) {
        400 => 'That request was not understood by the exhibition system.',
        401 => 'Please sign in again.',
        404 => 'That is no longer available.',
        429 => 'Too many attempts. Wait a moment and try again.',
        >= 500 => 'The exhibition system had a problem. Try again shortly.',
        _ => 'Something went wrong (error $status).',
      };

  Map<String, dynamic> _object(dynamic value) =>
      value is Map<String, dynamic> ? value : const {};

  // --- signing in ----------------------------------------------------------

  Future<LoginCodeRequest> requestLoginCode(String email) async {
    final json = await _send('POST', '/auth/request-code', body: {
      'email': email,
    });
    return LoginCodeRequest.fromJson(_object(json));
  }

  /// Exchanges the emailed code for a device token, which is set on this client
  /// immediately so the caller does not have to remember to.
  Future<({String token, Visitor visitor})> verifyLoginCode({
    required String email,
    required String code,
    String? platform,
    String? deviceName,
    String? appVersion,
  }) async {
    final json = _object(await _send('POST', '/auth/verify', body: {
      'email': email,
      'code': code,
      'platform': platform,
      'deviceName': deviceName,
      'appVersion': appVersion,
    }));

    final token = json['token']?.toString();
    if (token == null || token.isEmpty) {
      throw ApiException('The exhibition system did not return a sign-in.');
    }

    _token = token;
    return (token: token, visitor: Visitor.fromJson(_object(json['visitor'])));
  }

  Future<void> logout() async {
    try {
      await _send('POST', '/auth/logout');
    } on ApiException {
      // Signing out locally must succeed even when the server cannot be
      // reached; the token expires on its own regardless.
    }
    _token = null;
  }

  Future<Visitor> me() async => Visitor.fromJson(_object(await _send('GET', '/me')));

  Future<Visitor> updateConsent({bool? email, bool? tracking}) async =>
      Visitor.fromJson(_object(await _send('PATCH', '/me/consent', body: {
        'consentEmail': email,
        'consentTracking': tracking,
      })));

  // --- the exhibition ------------------------------------------------------

  /// One call at start-up: name, halls, the whole category tree, the countries
  /// and the programme's days. Everything the filter pickers need.
  Future<Exhibition> exhibition() async =>
      Exhibition.fromJson(_object(await _send('GET', '/exhibition')));

  Future<List<Hall>> halls() async {
    final json = await _send('GET', '/halls');
    return (json as List? ?? const [])
        .whereType<Map<String, dynamic>>()
        .map(Hall.fromJson)
        .toList();
  }

  Future<HallDetail> hall(int id, {int page = 1, int pageSize = 50}) async =>
      HallDetail.fromJson(_object(await _send('GET', '/halls/$id', query: {
        'page': page,
        'pageSize': pageSize,
      })));

  Future<List<Category>> categories() async {
    final json = await _send('GET', '/categories');
    return (json as List? ?? const [])
        .whereType<Map<String, dynamic>>()
        .map(Category.fromJson)
        .toList();
  }

  // --- exhibitors ----------------------------------------------------------

  Future<Paged<Exhibitor>> exhibitors({
    String? query,
    int? categoryId,
    int? subCategoryId,
    int? hallId,
    String? country,
    int page = 1,
    int pageSize = 25,
  }) async =>
      Paged.fromJson(
        _object(await _send('GET', '/exhibitors', query: {
          'q': query,
          'categoryId': categoryId,
          'subCategoryId': subCategoryId,
          'hallId': hallId,
          'country': country,
          'page': page,
          'pageSize': pageSize,
        })),
        Exhibitor.fromJson,
      );

  Future<ExhibitorDetail> exhibitor(int id) async =>
      ExhibitorDetail.fromJson(_object(await _send('GET', '/exhibitors/$id')));

  Future<SearchResults> searchEverything(String query) async =>
      SearchResults.fromJson(
          _object(await _send('GET', '/search', query: {'q': query})));

  // --- meetings and lectures -----------------------------------------------

  Future<Paged<Session>> sessions({
    String? query,
    DateTime? date,
    String? kind,
    int? hallId,
    int? categoryId,
    int? subCategoryId,
    bool bookmarkedOnly = false,
    int page = 1,
    int pageSize = 50,
  }) async =>
      Paged.fromJson(
        _object(await _send('GET', '/sessions', query: {
          'q': query,
          'date': date == null ? null : _dateParam(date),
          'kind': kind,
          'hallId': hallId,
          'categoryId': categoryId,
          'subCategoryId': subCategoryId,
          'bookmarked': bookmarkedOnly ? 'true' : null,
          'page': page,
          'pageSize': pageSize,
        })),
        Session.fromJson,
      );

  Future<SessionDetail> session(int id) async =>
      SessionDetail.fromJson(_object(await _send('GET', '/sessions/$id')));

  Future<void> setBookmarked(int sessionId, bool bookmarked) => _send(
        bookmarked ? 'POST' : 'DELETE',
        '/sessions/$sessionId/bookmark',
      );

  Future<Paged<Session>> agenda() async =>
      Paged.fromJson(_object(await _send('GET', '/me/agenda')), Session.fromJson);

  // --- e-catalogues --------------------------------------------------------

  /// Records a scanned stand QR code against the signed-in visitor. [scanned]
  /// is whatever the camera read — the server takes the token out of it, so the
  /// app does not have to know the URL shape.
  Future<ScanResult> scan(String scanned) async =>
      ScanResult.fromJson(_object(await _send('POST', '/me/scan', body: {
        'token': scanned,
      })));

  Future<List<ScannedStand>> myCatalogues() async {
    final json = _object(await _send('GET', '/me/catalogues'));
    return (json['items'] as List? ?? const [])
        .whereType<Map<String, dynamic>>()
        .map(ScannedStand.fromJson)
        .toList();
  }

  /// Requests a catalogue from an exhibitor's page instead of at the stand.
  /// Goes through the same path a scan does, so doing both does not put the
  /// exhibitor in the evening pack twice.
  Future<int> requestCatalogue(int kioskId) async {
    final json = _object(await _send('POST', '/me/catalogues', body: {
      'kioskId': kioskId,
    }));
    return _int(json['todayCount']);
  }

  Future<void> setCatalogueIncluded(int kioskId, bool included) => _send(
        'PATCH',
        '/me/catalogues/$kioskId',
        body: {'included': included},
      );

  static int _int(dynamic value) =>
      value is int ? value : int.tryParse(value?.toString() ?? '') ?? 0;

  // --- the visitor's day ---------------------------------------------------

  Future<VisitorDay> myDay() async =>
      VisitorDay.fromJson(_object(await _send('GET', '/me/day')));

  static String _dateParam(DateTime date) =>
      '${date.year.toString().padLeft(4, '0')}-'
      '${date.month.toString().padLeft(2, '0')}-'
      '${date.day.toString().padLeft(2, '0')}';

  void close() => _http.close();
}
